### PR TITLE
[NOD-721] Add defers

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: peer/peer.go [hunk 1]
```diff
@@ -494,8 +495,8 @@ func (p *Peer) StatsSnapshot() *StatsSnap {
 // This function is safe for concurrent access.
 func (p *Peer) ID() int32 {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	id := p.id
-	p.flagsMtx.Unlock()
 
 	return id
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 2]
```diff
@@ -505,8 +506,8 @@ func (p *Peer) ID() int32 {
 // This function is safe for concurrent access.
 func (p *Peer) NA() *wire.NetAddress {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	na := p.na
-	p.flagsMtx.Unlock()
 
 	return na
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 3]
```diff
@@ -532,8 +533,8 @@ func (p *Peer) Inbound() bool {
 // This function is safe for concurrent access.
 func (p *Peer) Services() wire.ServiceFlag {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	services := p.services
-	p.flagsMtx.Unlock()
 
 	return services
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 4]
```diff
@@ -543,17 +544,17 @@ func (p *Peer) Services() wire.ServiceFlag {
 // This function is safe for concurrent access.
 func (p *Peer) UserAgent() string {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	userAgent := p.userAgent
-	p.flagsMtx.Unlock()
 
 	return userAgent
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 5]
```diff
@@ -543,17 +544,17 @@ func (p *Peer) Services() wire.ServiceFlag {
...
 }
 
 // SubnetworkID returns peer subnetwork ID
 func (p *Peer) SubnetworkID() *subnetworkid.SubnetworkID {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	subnetworkID := p.cfg.SubnetworkID
-	p.flagsMtx.Unlock()
 
 	return subnetworkID
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 6]
```diff
@@ -563,8 +564,8 @@ func (p *Peer) SubnetworkID() *subnetworkid.SubnetworkID {
 // This function is safe for concurrent access.
 func (p *Peer) LastPingNonce() uint64 {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	lastPingNonce := p.lastPingNonce
-	p.statsMtx.RUnlock()
 
 	return lastPingNonce
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 7]
```diff
@@ -574,8 +575,8 @@ func (p *Peer) LastPingNonce() uint64 {
 // This function is safe for concurrent access.
 func (p *Peer) LastPingTime() time.Time {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	lastPingTime := p.lastPingTime
-	p.statsMtx.RUnlock()
 
 	return lastPingTime
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 8]
```diff
@@ -585,8 +586,8 @@ func (p *Peer) LastPingTime() time.Time {
 // This function is safe for concurrent access.
 func (p *Peer) LastPingMicros() int64 {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	lastPingMicros := p.lastPingMicros
-	p.statsMtx.RUnlock()
 
 	return lastPingMicros
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 9]
```diff
@@ -597,8 +598,8 @@ func (p *Peer) LastPingMicros() int64 {
 // This function is safe for concurrent access.
 func (p *Peer) VersionKnown() bool {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	versionKnown := p.versionKnown
-	p.flagsMtx.Unlock()
 
 	return versionKnown
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 10]
```diff
@@ -609,8 +610,8 @@ func (p *Peer) VersionKnown() bool {
 // This function is safe for concurrent access.
 func (p *Peer) VerAckReceived() bool {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	verAckReceived := p.verAckReceived
-	p.flagsMtx.Unlock()
 
 	return verAckReceived
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 11]
```diff
@@ -620,8 +621,8 @@ func (p *Peer) VerAckReceived() bool {
 // This function is safe for concurrent access.
 func (p *Peer) ProtocolVersion() uint32 {
 	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
 	protocolVersion := p.protocolVersion
-	p.flagsMtx.Unlock()
 
 	return protocolVersion
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 12]
```diff
@@ -631,8 +632,8 @@ func (p *Peer) ProtocolVersion() uint32 {
 // This function is safe for concurrent access.
 func (p *Peer) SelectedTipHash() *daghash.Hash {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	selectedTipHash := p.selectedTipHash
-	p.statsMtx.RUnlock()
 
 	return selectedTipHash
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 13]
```diff
@@ -683,8 +684,8 @@ func (p *Peer) BytesReceived() uint64 {
 // This function is safe for concurrent access.
 func (p *Peer) TimeConnected() time.Time {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	timeConnected := p.timeConnected
-	p.statsMtx.RUnlock()
 
 	return timeConnected
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: peer/peer.go [hunk 14]
```diff
@@ -696,8 +697,8 @@ func (p *Peer) TimeConnected() time.Time {
 // This function is safe for concurrent access.
 func (p *Peer) TimeOffset() int64 {
 	p.statsMtx.RLock()
+	defer p.statsMtx.RUnlock()
 	timeOffset := p.timeOffset
-	p.statsMtx.RUnlock()
 
 	return timeOffset
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please combine the above two lines.

File: rpcclient/infrastructure.go [hunk 15]
```diff
@@ -442,8 +442,8 @@ out:
 // is being reassigned during a reconnect.
 func (c *Client) disconnectChan() <-chan struct{} {
 	c.mtx.Lock()
+	defer c.mtx.Unlock()
 	ch := c.disconnect
-	c.mtx.Unlock()
 	return ch
```
Comment (stasatdaglabs) [inline rpcclient/infrastructure.go]: Please combine the above two lines.

File: rpcmodel/command_info.go [hunk 16]
```diff
@@ -10,15 +10,38 @@ import (
 	"strings"
 )
 
+func concreteTypeToMethodWithRLock(rt reflect.Type) (string, bool) {
+	registerLock.RLock()
+	defer registerLock.RUnlock()
+	method, ok := concreteTypeToMethod[rt]
+	return method, ok
```
Comment (stasatdaglabs) [inline rpcmodel/command_info.go]: Please combine the above two lines.

File: rpcmodel/command_info.go [hunk 17]
```diff
@@ -10,15 +10,38 @@ import (
...
+}
+
+func methodToInfoWithRLock(method string) (methodInfo, bool) {
+	registerLock.RLock()
+	defer registerLock.RUnlock()
+	info, ok := methodToInfo[method]
+	return info, ok
```
Comment (stasatdaglabs) [inline rpcmodel/command_info.go]: Please combine the above two lines.

File: server/p2p/p2p.go [hunk 18]
```diff
@@ -303,17 +303,17 @@ func (sp *Peer) addressKnown(na *wire.NetAddress) bool {
...
 // relayTxDisabled returns whether or not relaying of transactions for the given
 // peer is disabled.
 // It is safe for concurrent access.
 func (sp *Peer) relayTxDisabled() bool {
 	sp.relayMtx.Lock()
+	defer sp.relayMtx.Unlock()
 	isDisabled := sp.DisableRelayTx
-	sp.relayMtx.Unlock()
 
 	return isDisabled
```
Comment (stasatdaglabs) [inline server/p2p/p2p.go]: Please combine the above two lines.

File: server/rpc/rpcwebsocket.go [hunk 19]
```diff
@@ -1250,8 +1252,8 @@ func (c *wsClient) QueueNotification(marshalledJSON []byte) error {
 // Disconnected returns whether or not the websocket client is disconnected.
 func (c *wsClient) Disconnected() bool {
 	c.Lock()
+	defer c.Unlock()
 	isDisconnected := c.disconnected
-	c.Unlock()
 
 	return isDisconnected
```
Comment (stasatdaglabs) [inline server/rpc/rpcwebsocket.go]: Please combine the above two lines.

File: util/bloom/filter.go [hunk 20]
```diff
@@ -156,8 +156,8 @@ func (bf *Filter) matches(data []byte) bool {
 // This function is safe for concurrent access.
 func (bf *Filter) Matches(data []byte) bool {
 	bf.mtx.Lock()
+	defer bf.mtx.Unlock()
 	match := bf.matches(data)
-	bf.mtx.Unlock()
 	return match
```
Comment (stasatdaglabs) [inline util/bloom/filter.go]: Please combine the above two lines.

File: util/bloom/filter.go [hunk 21]
```diff
@@ -180,8 +180,8 @@ func (bf *Filter) matchesOutpoint(outpoint *wire.Outpoint) bool {
 // This function is safe for concurrent access.
 func (bf *Filter) MatchesOutpoint(outpoint *wire.Outpoint) bool {
 	bf.mtx.Lock()
+	defer bf.mtx.Unlock()
 	match := bf.matchesOutpoint(outpoint)
-	bf.mtx.Unlock()
 	return match
```
Comment (stasatdaglabs) [inline util/bloom/filter.go]: Please combine the above two lines.

File: util/bloom/filter.go [hunk 22]
```diff
@@ -330,8 +330,8 @@ func (bf *Filter) matchTxAndUpdate(tx *util.Tx) bool {
 // This function is safe for concurrent access.
 func (bf *Filter) MatchTxAndUpdate(tx *util.Tx) bool {
 	bf.mtx.Lock()
+	defer bf.mtx.Unlock()
 	match := bf.matchTxAndUpdate(tx)
-	bf.mtx.Unlock()
 	return match
```
Comment (stasatdaglabs) [inline util/bloom/filter.go]: Please combine the above two lines.

File: util/bloom/filter.go [hunk 23]
```diff
@@ -341,7 +341,7 @@ func (bf *Filter) MatchTxAndUpdate(tx *util.Tx) bool {
 // This function is safe for concurrent access.
 func (bf *Filter) MsgFilterLoad() *wire.MsgFilterLoad {
 	bf.mtx.Lock()
+	defer bf.mtx.Unlock()
 	msg := bf.msgFilterLoad
-	bf.mtx.Unlock()
 	return msg
```
Comment (stasatdaglabs) [inline util/bloom/filter.go]: Please combine the above two lines.

File: server/rpc/rpcwebsocket.go [hunk 24]
```diff
@@ -361,6 +361,12 @@ func (f *wsClientFilter) existsUnspentOutpoint(op *wire.Outpoint) bool {
 	return ok
 }
 
+func (f *wsClientFilter) existsUnspentOutpointWithLock(op *wire.Outpoint) bool {
```
Comment (svarogg) [inline server/rpc/rpcwebsocket.go]: We agreed some time ago that the standard for functions with that have lock/no-lock version is xxx and xxxNoLock.
Lets keep it this way.

File: server/rpc/rpcwebsocket.go [hunk 25]
```diff
@@ -1289,6 +1289,12 @@ func (c *wsClient) WaitForShutdown() {
 	c.wg.Wait()
 }
 
+func (c *wsClient) filterDataWithLock() *wsClientFilter {
```
Comment (svarogg) [inline server/rpc/rpcwebsocket.go]: Here too

File: server/rpc/handle_get_block_template.go [hunk 26]
```diff
@@ -217,26 +217,73 @@ func handleGetBlockTemplateRequest(s *Server, request *rpcmodel.TemplateRequest,
...
+	result, err = state.blockTemplateResult(s.cfg.DAG, useCoinbaseValue)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// blockTemplateOrLongPollChan returns a block template if the
+// template identified by the provided long poll ID is stael or
```
Comment (stasatdaglabs) [inline server/rpc/handle_get_block_template.go]: Tyop: stael -> stale

File: server/rpc/handle_get_block_template.go [hunk 27]
```diff
@@ -217,26 +217,73 @@ func handleGetBlockTemplateRequest(s *Server, request *rpcmodel.TemplateRequest,
...
+		return result, nil
+	}
+
+	select {
+	// When the client closes before it's time to send a reply, just return
+	// now so the goroutine doesn't hang around.
+	case <-closeChan:
+		return nil, ErrClientQuit
+
+		// Wait until signal received to send the reply.
```
Comment (stasatdaglabs) [inline server/rpc/handle_get_block_template.go]: Bad indentation. Please fix.

**Pull request merged**